keps/498-GroupScheduling/README.md

```diff
@@ -0,0 +1,336 @@
+  groupSchedulingConfig:
+    targetReplicatedJobs:
+    - workers
+  replicatedJobs:
+    - name: workers
+      replicas: 3
+      template:
+        spec:
+          backoffLimit: 0
+          completions: 100
+          parallelism: 100
+          template:
+            spec:
+              containers:
+                - name: pytorch
+                  image: pytorch/pytorch:latest
+                  command:
+                    - bash
+                    - -c
+                    - train.py
+```
+
+### Constraints
+
+In Order Startup Policy is incompatible with a group scheduling config that spans
+multiple replicated jobs, since they are conceptually mutually exclusive.
+
+We must add validation to our JobSet webhook to enforce this.
+
+### Implmentation
+
+
+A new pod controller will be added: `group_scheduling_pod_controller.go`, which
+reconciles on leader pods which are scheduled and whose JobSet have a group scheduling
+config defined.
+
+When a GroupSchedulingConfig is defined:
+
+ 1. A bash initContainer will be injected into the JobTemplateSpec of every 
+    TargetReplicatedJob by the existing pod mutating webhook. 
+    The initContainer will have a simple bash loop, checking for a particular
+    directory path to exist. Once the path exists, it will exit successfully.
+ 2. A shared ConfigMap will be created and mounted into the initContainer as an 
+    optional volume mount. The ConfigMap will be empty to start, so the mount 
+    path will not exist in the initContainer yet.
+ 3. The pod controller will reconcile on leader pods, listing and
+    counting the number of pods which have started in the JobSet. If the started 
+    count is less than the expected count, re-enqueue the pod for reconciliation 
+    after a short interval.
+ 4. Once the expected number of pods for the JobSet have started (derived from the
+    JobSet spec), the pod controller will update the ConfigMap with some arbitrary data
+    so that it is no longer empty. This will trigger Kubelet to populate
+    the initContainer mount point. 
+ 5. Now that the mount point directory path exists in the initContainer, it
+    will exit and allow the main workload container to start.
+
+
+#### Example JobSet spec AFTER webhook injection:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: my-jobset
+  namespace: default
+spec:
+  # main containers in pods part of `workers` replicatedJob will be blocked
+  # from execution until all pods have started.
+  groupSchedulingConfig:
+    targetReplicatedJobs:
+    - workers
+  replicatedJobs:
+    - name: workers
+      replicas: 3
+      template:
+        spec:
+          backoffLimit: 0
+          completions: 100
+          parallelism: 100
+          template:
+            spec:
+              initContainers:
+                # Check if the started file is present before exiting and starting the main container.
+                - name: group-scheduling-init
```

Comment:
We probably should mention what names will be forbidden for the "main pod"?

I'm not sure if the name is taking by the main container, will the init container take perference over the main container? Or will they have conflicts.

Reply:
Good point, I will add something to "constraints" section about this. We can validate this in the JobSet webhook to ensure there's no conflicts.